Issue 11 Setting dimlevel wont work

kskaret opened the pull request:
Fix to https://github.com/amlinger/homebridge-telldus-tdtool/issues/11

amlinger committed:
Should fix issue#7, Siri latency when turning off and on lights.

kskaret committed:
replacing the endswith('Success') logic which doesnt seem to work

kskaret committed:
correcting dimming not working. Sending --dimlevel and --dim in same command

kskaret committed:
Ajusting dimlevel from percentage into 0-255 value

Commit:
Remove temp logging

amlinger commented on src/lib/tdtool.js:
```diff
@@ -48,6 +48,9 @@ const TDtool = {
   run: (cmd, target) => TDtool.isInstalled().then(() =>
     execute(`tdtool ${cmd} ${target}`)),
 
+ run: (cmd1, target1, cmd2, target2) => TDtool.isInstalled().then(() =>
```
This actually overrides the previous `run` on line 48, which might/will cause unexpected behaviour.

amlinger commented on src/lib/tdtool.js:
```diff
@@ -59,7 +62,7 @@ const TDtool = {
   off: id => TDtool.run('--off', id),
   dim: id => TDtool.run('--dim', id),
-  dimlevel: level => TDtool.run('--dimlevel', level),
+  dimlevel: (level, id) => TDtool.run('--dimlevel', level, '--dim', id),
```
A better option to me would be removing `dimlevel`, and implementing `dim` with the `execute`-implementation:

``` js
dim: (level, id) => TDtool.isInstalled().then(() =>
      execute(`tdtool --dimlevel ${level} --dim ${id}`)),
```

This will keep the original `run` implementation and not break the existing actions.

amlinger commented on src/lib/telldus-accessory.js:
```diff
@@ -74,12 +74,19 @@ class TelldusAccessory {
       case this.Characteristic.Formats.INT:
-        TDtool.setDimLevel(value).then(() => TDtool.dim(this.id)).then(out => {
-          callback(null, out.endswith('Success'))
+        TDtool.dimlevel(value*255.0/100, this.id).then(out => {
```
This module actually already contains a method for converting this value, look at https://github.com/amlinger/homebridge-telldus-tdtool/blob/master/src/lib/telldus-accessory.js#L8 and see if that suits your needs :)

amlinger commented:
Hi @kskaret, 

Thanks for contributing, its really helpful since I do not have any dimmers myself yet :)

It appears as you have not based this branch on master, but on #7, which makes some commits appear in this branch that should not be here, and will cause trouble in a later merge. Please include only the commit you would like to merge (preferably only one).

Commit:
Removing the extra run method and reimplemeting the dim method

kskaret committed:
Using percentageToBits function

kskaret commented:
Hi @amlinger 
Thanks for the feedback. I did some commits regarded the things you pointed out.
As you may notice, I dont have any node or javascript experience at all before this (I do mostly Java). But this is the best way to learn!
I'm happy to contribute, because I'm just eager to make this work perfectly in my house.

Btw, I still experience another small problem, but I think I see whats wrong. So you can expect something more from my side when I have more time to look into it.

amlinger commented:
Looks good to me!

I am just a bit concerned at the moment about the git history (with the history containing many commits, and the branch not being based on `master`, but on `issue-7/fix-siri-latency`. These are things that might come back and bite you, since 
- Rolling back is not easy with the feature spread over many commits.
- Merging the other branch will be a lot harder, if that were to diverge further.

A fix could be to rebase your branch on master:

``` bash
git rebase -i master
```

and then ignoring the commits related to the other branch, and squashing the rest: 

``` git
# pick 3cd380a Should fix issue#7, Siri latency when turning off and on lights.
# pick 4058ab1 replacing the endswith('Success') logic which doesnt seem to work
pick correcting dimming not working. Sending --dimlevel and --dim in same command
squash f95a9c3 Ajusting dimlevel from percentage into 0-255 value
squash 12fa60b Remove temp logging
squash 6dfb991 Removing the extra run method and reimplemeting the dim method
squash 0f507db Using percentageToBits function
```

This will of course cause conflicts along the way, which will need to be resolved in each step, keeping the old callback implementation as the other branch is dealing with it. 

In the end, you should end up with something like `git diff master`:

```
diff --git a/src/lib/tdtool.js b/src/lib/tdtool.js
index cea61b6..a3173f7 100644
--- a/src/lib/tdtool.js
+++ b/src/lib/tdtool.js
@@ -58,8 +58,8 @@ const TDtool = {
    */
   on:  id => TDtool.run('--on', id),
   off: id => TDtool.run('--off', id),
-  dim: id => TDtool.run('--dim', id),
-  dimlevel: level => TDtool.run('--dimlevel', level),
+  dim: (level, id) => TDtool.isInstalled().then(() =>
+      execute(`tdtool --dimlevel ${level} --dim ${id}`)),
 }

 module.exports = TDtool
diff --git a/src/lib/telldus-accessory.js b/src/lib/telldus-accessory.js
index 30bf75d..3c01f02 100644
--- a/src/lib/telldus-accessory.js
+++ b/src/lib/telldus-accessory.js
@@ -78,7 +78,7 @@ class TelldusAccessory {
         }, error => callback(new Error(error)))
         break
       case this.Characteristic.Formats.INT:
-        TDtool.setDimLevel(value).then(() => TDtool.dim(this.id)).then(out => {
+        TDtool.dim(percentageToBits(value), this.id).then(out => {
           callback(null, out.endswith('Success'))
         }, error => callback(new Error(error)))
         break
```

and `git log` should say that you've only added one commit:) 

Regarding the other problem, feel free to open another issue, you never know whom else it might help out :)

kskaret commented:
Hello! I created another pull request, squashed and rebased as you wished :)
https://github.com/amlinger/homebridge-telldus-tdtool/pull/14

I agree it looks cleaner now, althought wheater or not to squash commits is a matter of taste. I've seen projects doing different styles, and I'm actually not used to doing it my self. But interesting to try out

amlinger commented:
Closing this in favour of #14.
